vm/src/stdlib/itertools.rs

```diff
@@ -1185,21 +1185,31 @@ mod decl {
     struct PyItertoolsTeeData {
         iterable: PyIter,
         values: PyRwLock<Vec<PyObjectRef>>,
+        locked: AtomicCell<bool>,
     }
 
     impl PyItertoolsTeeData {
         fn new(iterable: PyIter, _vm: &VirtualMachine) -> PyResult<PyRc<Self>> {
             Ok(PyRc::new(Self {
                 iterable,
                 values: PyRwLock::new(vec![]),
+                locked: AtomicCell::new(false),
             }))
         }
 
         fn get_item(&self, vm: &VirtualMachine, index: usize) -> PyResult<PyIterReturn> {
             if self.values.read().len() == index {
-                let result = raise_if_stop!(self.iterable.next(vm)?);
-                self.values.write().push(result);
+                if self.locked.swap(true) {
+                    return Err(vm.new_runtime_error("cannot re-enter the tee iterator"));
+                }
+
+                let result = self.iterable.next(vm);
+                self.locked.store(false);
+
+                let obj = raise_if_stop!(result?);
+                self.values.write().push(obj);
             }
+
```

Comment:
While traditional locks prefer to lock the code, Rust locks prefer to lock the data.
After changing PyRwLock to PyMutex, this approach will work
```suggestion
            let Some(values) = self.values.try_lock() else {
                return Err(vm.new_runtime_error("cannot re-enter the tee iterator"));
            };
            if values.len() == index {
                let obj = raise_if_stop!(self.iterable.next(vm)?);
                values.push(obj);
            }
            Ok(PyIterReturn::Return(values[index].clone()))
```